File: 2019/day8/day8.py
#!/usr/bin/env python3
# [Day 8: Space Image Format](https://adventofcode.com/2019/day/8)

import sys
from pathlib import Path

sys.path.append(Path(__file__).parent.parent.as_posix())
from ocr.ocr import ocr  # noqa


def chunker(seq, size):
    return (seq[pos : pos + size] for pos in range(0, len(seq), size))


filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()


# part 1
layers = []
for i, layer in enumerate(chunker(data, 25 * 6), 1):
    layers.append((layer.count("0"), layer.count("1") * layer.count("2")))
print(sorted(layers)[0][1])

# part 2
w, h = 25, 6
image = [[" "] * w for _ in range(h)]
for y in range(h):
    for x in range(w):
        for i in range(len(data) // (w * h)):
            pixel = data[(x + y * w) + i * w * h]
            if pixel == "2":
                continue
            if pixel == "1":
                image[h - 1 - y][x] = "#"
            if pixel == "0":
                image[h - 1 - y][x] = "."
            break
password = "\n".join("".join(row) for row in reversed(image))
print(ocr(password))

File: 2019/day8/day8.rs
//! [Day 8: Space Image Format](https://adventofcode.com/2019/day/8)

struct Puzzle {
    data: Vec<u8>,
}

impl Puzzle {
    /// Initialize from the puzzle input.
    fn new(data: &str) -> Self {
        Self {
            data: data.trim_ascii().bytes().collect(),
        }
    }

    /// Solve part one.
    fn part1(&self) -> i32 {
        let mut layers = Vec::new();

        for layer in self.data.chunks(25 * 6) {
            let mut zeros = 0;
            let mut ones = 0;
            let mut twos = 0;

            for c in layer {
                match c {
                    b'0' => zeros += 1,
                    b'1' => ones += 1,
                    b'2' => twos += 1,
                    _ => {}
                }
            }

            layers.push((zeros, ones * twos));
        }

        layers.sort_unstable();

        layers[0].1
    }

    fn make_image(&self, w: usize, h: usize) -> String {
        let mut image = aoc::GridU::<char>::with_size(w, h);

        for ((x, y), c) in image.iter_mut() {
            for i in 0..(self.data.len() / (w * h)) {
                let pixel = self.data[(x + y * w) + i * w * h];
                match pixel {
                    b'0' => {
                        *c = '.';
                        break;
                    }
                    b'1' => {
                        *c = '#';
                        break;
                    }
                    _ => {}
                }
            }
        }

        format!("{image}")
    }

    /// Solve part two.
    fn part2(&self) -> String {
        aoc::ocr::scan_5x6(&self.make_image(25, 6))
    }
}

fn main() {
    let args = aoc::parse_args();
    let puzzle = Puzzle::new(&args.input);
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn part2() {
        let puzzle = Puzzle::new("0222112222120000");
        assert_eq!(puzzle.make_image(2, 2), ".#\n#.\n");
    }
}
